example/socks/http.cpp

```diff
@@ -0,0 +1,355 @@
+{
+    std::cerr << what << ": " << ec.message() << "\n";
+}
+
+// Performs an HTTP GET and prints the response
+void
+do_session(
+    std::string const& host,
+    std::string const& port,
+    std::string const& target,
+    int version,
+    std::string const& socks_server,
+    boost::asio::io_context& ioc,
+    boost::asio::yield_context yield)
+{
+    boost::system::error_code ec;
+
+    // These objects perform our I/O
+    tcp::resolver resolver{ioc};
+    tcp::socket socket{ioc};
+
+    socks_url url;
+    if (!parse_socks_url(socks_server, url))
+        return fail(ec, "parse_socks_url");
+
+    // socks handshake.
+    int socks_version = url.scheme == "socks4" ? 4 : 5;
+    socks_version = url.scheme == "socks5" ? 5 : 0;
+    if (socks_version == 0)
+        return fail(ec, "incorrect socks version");
+
+    // Look up the domain name
+    auto const results = resolver.async_resolve(url.host, url.port, yield[ec]);
+    if (ec)
+        return fail(ec, "resolve");
+
+    // Make the connection on the IP address we get from a lookup
+    boost::asio::async_connect(socket, results.begin(), results.end(), yield[ec]);
+    if (ec)
+        return fail(ec, "connect");
+
+    socks::async_handshake(socket, host, static_cast<unsigned short>(std::atoi(port.c_str())),
+        socks_version, url.username, url.password, true, yield[ec]);
+    if (ec)
+        return fail(ec, "socks async_handshake");
+
+    // Set up an HTTP GET request message
+    http::request<http::string_body> req{http::verb::get, target, version};
+    req.set(http::field::host, host);
+    req.set(http::field::user_agent, BOOST_BEAST_VERSION_STRING);
+
+    // Send the HTTP request to the remote host
+    http::async_write(socket, req, yield[ec]);
+    if(ec)
+        return fail(ec, "write");
+
+    // This buffer is used for reading and must be persisted
+    boost::beast::flat_buffer b;
+
+    // Declare a container to hold the response
+    http::response<http::dynamic_body> res;
+
+    // Receive the HTTP response
+    http::async_read(socket, b, res, yield[ec]);
+    if(ec)
+        return fail(ec, "read");
+
+    // Write the message to standard out
+    std::cout << res << std::endl;
+
+    // Gracefully close the socket
+    socket.shutdown(tcp::socket::shutdown_both, ec);
+
+    // not_connected happens sometimes
+    // so don't bother reporting it.
+    //
+    if(ec && ec != boost::system::errc::not_connected)
+        return fail(ec, "shutdown");
+
+    // If we get here then the connection is closed gracefully
+}
+
+//------------------------------------------------------------------------------
+
+int main(int argc, char** argv)
+{
+    // Check command line arguments.
+    if(argc != 5)
+    {
+        std::cerr <<
+            "Usage: socsk <host> <port> <target> <socks[4|5]://[user[:password]@]server:port>\n" <<
+            "Example:\n" <<
+            "    socsk www.example.com 80 / socks5://socks5server.com:1080\n";
```

Comment:
The example line is wrong. "socsk" is misspelled. I think this program should be renamed to "socks-client-coro" and the source file renamed to socks_client_coro.cpp